docs: snowflake_plus: document `iceberg_mode` and `base_location` templating (#2524)
dlt-hub/dlt: docs/website/docs/plus/ecosystem/snowflake_plus.md

  </TabItem>
</Tabs>

## Configuration

The `snowflake_plus` destination extends the standard Snowflake configuration with additional options:

| Option | Description | Required | Default |
|--------|-------------|----------|---------|
| `iceberg_mode` | Controls which tables are created as Iceberg tables. Possible values: <ul><li>`"all"`: All tables including dlt system tables are created as Iceberg tables</li><li>`"data_tables"`: Only data tables (non-dlt system tables) are created as Iceberg tables</li><li>`"none"`: No tables are created as Iceberg tables</li></ul> | No | `"none"` |
| `external_volume` | The external volume to store Iceberg metadata | Yes | None |
| `catalog` | The catalog to use for Iceberg tables | No | `"SNOWFLAKE"` |

> | `base_location` | Template string for the base location where Iceberg data is stored in the external volume. Supports placeholders like `{dataset_name}` and `{table_name}`. | No | `"{dataset_name}/{table_name}"` |

Reviewer: Idk why this quotation mark is on a separate line, probably because of table width? 
<img width="934" alt="image" src="https://github.com/user-attachments/assets/98a9dafa-cd59-4978-847b-71ed1394829e" />

This also doesn't look the best in a table, not sure if we can do much about that though 🤔 
<img width="926" alt="image" src="https://github.com/user-attachments/assets/c7cb5d78-bc94-49cf-8115-12ae34629f2c" />

Author: good point, so far I haven't found an easy way to fix the wrapping issue with the quote. Let me try to replace table with something simpler.

Author: @akelad I've changed the table layout to a plain list of parameters in 68f19ec
Please take a look.

| `extra_placeholders` | Dictionary of additional values that can be used in the `base_location` template. The values can be static strings or functions that accept the dataset name and table name as arguments and return a string. | No | None |
| `catalog_sync` | The name of a [catalog integration](https://docs.snowflake.com/en/user-guide/tables-iceberg#catalog-integration) for syncing Iceberg tables to an external catalog in [Snowflake Open Catalog](https://other-docs.snowflake.com/en/opencatalog/overview) | No | None |

Configure these options in your `config.toml` file under the `[destination.snowflake]` section or in `dlt.yml` file under the `destinations.snowflake_plus` section.

## Base location templating

The `base_location` parameter controls where Snowflake stores your Iceberg table data and metadata in the external volume. It's a template string that supports the following built-in placeholders:

- `{dataset_name}`: The name of your dataset
- `{table_name}`: The name of the table

For more flexibility, you can also define custom placeholders using the `extra_placeholders` option.

### Examples

> 1. Default pattern (if not specified, `{dataset_name}/{table_name}`) creates paths like `my_dataset/customers` in your external volume.

Reviewer: ```suggestion
1. The default pattern `{dataset_name}/{table_name}` creates paths like `my_dataset/customers` in your external volume.
```

2. Custom static path:
   ```yaml
   base_location: "custom/static/path"
   ```
   This creates all tables in the same directory `custom/static/path`.

3. Using custom placeholders:
   ```yaml
   base_location: "{env}/{dataset_name}/{table_name}"
   extra_placeholders:
     env: "prod"
   ```
   This creates paths like `prod/my_dataset/customers`.

### How Snowflake uses the base location

When you provide a `base_location`, Snowflake uses it to create the paths where data and metadata are stored in your external cloud storage. The actual directory structure Snowflake creates follows this pattern:

```text
STORAGE_BASE_URL/BASE_LOCATION.<randomId>/[data | metadata]/
```

> Where `<randomId>` is a random, Snowflake-generated 8-character string appended to create a unique directory.

Reviewer: ```suggestion
Where `<randomId>` is a random Snowflake-generated 8-character string appended to create a unique directory.
```
